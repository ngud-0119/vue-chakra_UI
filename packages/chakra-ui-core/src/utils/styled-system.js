
/**
 * Existential getter function that can be used in any style declaration to get a value
 * from theme, with support for fallback values. This helps prevent errors from
 * throwing when a theme value is missing.
 * @param {Object} obj Theme property
 * @param {String} key Theme key
 * @param {String} def Definition if non-existent
 * @param {*} p
 * @param {*} undef
 */
export const __get = (obj, key, def, p, undef) => {
  key = key && key.split ? key.split('.') : [key]

  for (p = 0; p < key.length; p++) {
    obj = obj ? obj[key[p]] : undef
  }

  return obj === undef ? def : obj
}
