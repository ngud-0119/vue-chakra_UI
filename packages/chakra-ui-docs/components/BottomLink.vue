<template>
  <div>
    <CFlex justify="space-between">
      <div>
        <CLink as="router-link" :to="prev">
          {{ prev == '' ? "" : "Prev" }}
        </CLink>
      </div>

      <CLink as="router-link" :to="next">
        {{ next == '' ? "": "Next" }}
      </CLink>
    </CFlex>
  </div>
</template>

<script>
import { CLink, CFlex } from '@chakra-ui/vue'
import { findNextAndPrevRoute } from '../utils'

export default {
  components: {
    CLink,
    CFlex
  },
  data: () => ({
    prev: '',
    next: ''
  }),
  watch: {
    $route (to, from) {
      const { prev, next } = findNextAndPrevRoute(to.path, this.$router.options.routes)

      this.prev = prev
      this.next = next
    }
  },
  created () {
    const { prev, next } = findNextAndPrevRoute(this.$route.path, this.$router.options.routes)

    this.prev = prev
    this.next = next
  }
}
</script>

<style></style>
