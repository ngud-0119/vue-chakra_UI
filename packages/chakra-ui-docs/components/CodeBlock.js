import { CBox, CButton } from '@chakra-ui/vue'
import PrismEditor from 'vue-prism-editor'

export default {
  name: 'CodeBlock',
  props: {
    lang: {
      type: String,
      default: 'vue'
    },
    isReadOnly: {
      type: Boolean,
      default: true
    },
    emitEvents: {
      type: Boolean,
      default: false
    },
    lineNumbers: {
      type: Boolean,
      default: false
    },
    autoStyleLineNumbers: {
      type: Boolean,
      default: true
    },
    code: String
  },
  data () {
    return {
      text: undefined,
      copyTimeout: null,
      copyButtonText: 'Copy'
    }
  },
  methods: {
    async copy () {
      // Copy text to clipboard
      await navigator.clipboard.writeText(this.text)

      // Handle timeouts for copy button text
      if (this.copyTimeout) clearTimeout(this.copyTimeout)
      this.copyButtonText = 'Copied!'
      this.copyTimeout = setTimeout(() => {
        this.copyButtonText = 'Copy'
        clearTimeout(this.copyTimeout)
      }, 1000)
    }
  },
  render (h) {
    // const children = this.$slots.default[0]
    const innerText = this.code.trim()
    this.text = innerText

    return h(CBox, {
      props: {
        rounded: 'md',
        position: 'relative',
        fontSize: '0.9rem'
      }
    }, [
      h(PrismEditor, {
        props: {
          code: innerText,
          language: this.lang,
          readonly: this.isReadOnly,
          ...this.$props
        }
      }),
      h(CButton, {
        props: {
          variantColor: 'vue',
          position: 'absolute',
          size: 'sm',
          top: '0.2rem',
          right: '0.125rem',
          textTransform: 'uppercase',
          transform: 'scale(0.8)'
        },
        on: {
          click: this.copy
        }
      }, this.copyButtonText)
    ])
  }
}
