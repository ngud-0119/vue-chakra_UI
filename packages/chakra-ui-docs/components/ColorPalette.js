import { CBox, CFlex, CGrid, boxProps } from '@chakra-ui/vue'

export const ColorPalette = {
  props: {
    ...boxProps,
    color: String,
    name: String
  },
  inject: ['$chakraTheme'],
  computed: {
    theme () {
      return this.$chakraTheme()
    },
    colorCode () {
      let colorCode = this.color
      const [shade, hue] = this.color.split('.')

      if (shade && hue) {
        colorCode = this.theme.colors[shade][hue]
      }

      if (this.color in this.theme.colors && typeof this.theme.colors[this.color] === 'string') {
        colorCode = this.theme.colors[this.color]
      }

      return colorCode
    }
  },
  render (h) {
    return (
      <CFlex align="center" {...this.$props}>
        <CBox rounded="md" w="3rem" h="3rem" boxShadow="inner" mr={3} bg={this.color} />
        <CBox fontSize="sm">
          <CBox fontWeight="semibold" textTransform="capitalize">
            {this.name}
          </CBox>
          <CBox textTransform="uppercase">{this.colorCode}</CBox>
        </CBox>
      </CFlex>
    )
  }
}

export const ColorPalettes = {
  name: 'ColorPalettes',
  inject: ['$chakraTheme'],
  props: {
    color: String
  },
  computed: {
    theme () {
      return this.$chakraTheme()
    }
  },
  render (h) {
    const keys = Object.keys(this.theme.colors[this.color])
    return (
      <CGrid
        mt={7}
        gap={6}
        templateColumns="repeat( auto-fit, minmax(200px, 1fr) )"
        {...this.$attrs}
      >
        {keys.map(item => (
          <ColorPalette color={`${this.color}.${item}`} name={`${this.color} ${item}`} />
        ))}
      </CGrid>
    )
  }
}

export const ColorWrapper = {
  render (h) {
    return (
      <CGrid
        mt={7}
        gap={6}
        templateColumns="repeat( auto-fit, minmax(200px, 1fr) )"
        {...this.$attrs}
      >
        {this.$slots.default}
      </CGrid>
    )
  }
}
