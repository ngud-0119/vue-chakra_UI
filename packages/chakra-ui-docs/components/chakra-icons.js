export { default as chakraIcons } from '../../chakra-ui-core/src/lib/internal-icons'
