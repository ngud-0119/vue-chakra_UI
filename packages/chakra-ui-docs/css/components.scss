:root {
  --font-body: -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  --font-mono: SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
}

// h1 {
//   font-size: 1.875rem;
//   line-height: 1.25;
//   font-weight: 700;
//   font-family: var(--font-body);
//   margin-top: 1em;
//   margin-bottom: 1em;
// }

// h3 {
//   font-weight: 700;
//   font-size: 1rem;
// }

pre {
  margin-top: 2em;
  margin-bottom: 2em;
  border-radius: 4px;
}

table {
  text-align: left;
  margin-top: 32px;
  width: 100%;
}

th {
  background-color: #F7FAFC;
  font-weight: 500;
  font-size: 0.875rem;
  padding: 0.5rem;
}

td {
  padding: 0.5rem;
  border-top-width: 1px;
  border-color: inherit;
  font-size: 0.875rem;
  white-space: normal;
}

a {
  color: #55c392;
  cursor: pointer;
  text-decoration: underline;
  outline: none;

  &::hover {
    opacity: 0.8;
  }

  &::focus {
    box-shadow: 0 0 0 3px rgba(66, 153, 225, 0.6);
  }
}

code {
  font-family: var(--font-mono);
}

br {
  height: 24px;
}

p {
  margin-top: 1rem;
  line-height: 1.625;

  &:not(.preview-panel p) {
    margin-top: 0;
  }
}

ul, ol {
  padding-top: 8px;
  padding-left: 16px;
}

li {
  padding-bottom: 4px;
}

blockquote {
  display: flex;
  align-items: center;
  position: relative;
  padding-left: 0.75rem;
  padding-right: 1rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  background-color: rgb(254, 235, 200);
  margin-top: 1rem;
  overflow: hidden;
  border-left: 4px solid rgb(221, 107, 32);
  border-color: rgb(221, 107, 32);
  color: black;

  > *:first-of-type {
    margin-top: 0;
    color: inherit;
  }
}

@media (min-width: 768px) {
  h1 {
    font-size: 2.25rem;
  }

  h2 {
    font-size: 1.5rem;
  }

  h3 {
    font-size: 1.25rem
  }
}