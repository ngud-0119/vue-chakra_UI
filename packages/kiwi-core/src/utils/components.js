export const isVueComponent = (value) => {
  return (!!value && !!value.$el)
}
