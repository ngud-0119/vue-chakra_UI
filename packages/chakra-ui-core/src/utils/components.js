import { css } from '@emotion/css'
import { runIfFn } from '@chakra-ui/utils'
import { css as _css } from '@chakra-ui/styled-system'
import { __get } from './styled-system'
import { hasOwn, extractChakraAttrs } from './object'

export const isVueComponent = (value) => {
  return !!value && !!value.$el
}

/**
 * Makes a cache watcher handler for data property.
 * This utility helps prevent unnecessary re-renders
 * for primitives with changes in the $parent $attrs
 * and $listeners objects
 * @param {String} property
 */
export function createWatcher (property) {
  return {
    handler (newVal, oldVal) {
      for (const key in oldVal) {
        if (!hasOwn(newVal, key)) {
          this.$delete(this.$data[property], key)
        }
      }
      for (const key in newVal) {
        this.$set(this.$data[property], key, newVal[key])
      }
    },
    immediate: true
  }
}

/**
 * Create mixin for style attributes
 * @param {String} name Component name
 */
export const createStyledAttrsMixin = name => ({
  name,
  inheritAttrs: false,
  inject: ['$chakraTheme', '$chakraColorMode'],
  data () {
    return {
      attrs$: {},
      listeners$: {}
    }
  },
  props: {
    as: {
      type: [String, Object],
      default: 'div'
    },
    to: String
  },
  computed: {
    colorMode () {
      return this.$chakraColorMode()
    },
    theme () {
      return this.$chakraTheme()
    },
    /** Split style attributes and native attributes */
    splitProps () {
      const $attrs = this.$data.attrs$
      const { styleAttrs, nativeAttrs } = extractChakraAttrs($attrs)

      return {
        styleAttrs,
        nativeAttrs
      }
    },
    baseStyle () {
      const componentBaseStyleObjectOrFunction = __get(this.theme, `baseStyles.${name}`)
      return componentBaseStyleObjectOrFunction ? (
        runIfFn(componentBaseStyleObjectOrFunction, {
          ...this.splitProps.styleAttrs,
          ...this.splitProps.nativeAttrs,
          theme: this.theme,
          colorMode: this.colorMode
        })
      ) : {}
    },
    className () {
      const { styleAttrs } = this.splitProps
      const merged = {
        ...this.componentStyles || {},
        ...this.baseStyle,
        ...styleAttrs
      }
      const boxStylesObject = _css(merged)(this.theme)

      return css(boxStylesObject)
    },
    /** Computed attributes object */
    computedAttrs () {
      return {
        ...(name && { 'data-chakra-component': name }),
        ...this.splitProps.nativeAttrs
      }
    },
    /** Computed listeners object */
    computedListeners () {
      return this.$data.listeners$
    }
  },
  created () {
    this.$watch('$attrs', createWatcher('attrs$'))
    this.$watch('$listeners', createWatcher('listeners$'))
  }
})
