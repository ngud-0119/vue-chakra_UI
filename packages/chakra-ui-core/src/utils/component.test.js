import defaultTheme from '@chakra-ui/theme-vue'
import { toCSSVar } from '@chakra-ui/styled-system'
import { createStyledAttrsMixin } from './components'
import { render } from '@/tests/test-utils'

describe('createStyledAttrsMixin', () => {
  const FakeComponent = {
    name: 'FakeComponent',
    mixins: [createStyledAttrsMixin('FakeComponent', true)],
    render (h) {
      return h(
        'div',
        {
          class: this.className,
          attrs: this.computedAttrs
        },
        'Fake component'
      )
    }
  }

  describe('baseStyles', () => {
    const renderFakeComponent = ({ theme, colorMode, ...props }) => {
      const inlineAttrs = (props && props.inlineAttrs) || ''
      return render({
        template: `<FakeComponent ${inlineAttrs} />`,
        components: {
          FakeComponent
        },
        provide: () => ({
          $chakraTheme: () => toCSSVar(theme),
          $chakraColorMode: () => colorMode || 'light'
        }),
        ...props
      })
    }

    it('should use theme.baseStyles if given', () => {
      const { asFragment } = renderFakeComponent({
        theme: {
          ...defaultTheme,
          baseStyles: {
            FakeComponent: {
              bg: 'red.400',
              color: 'gray.200'
            }
          }
        }
      })
      expect(asFragment()).toMatchSnapshot()
    })

    it('should be overiden by props', () => {
      const { asFragment } = renderFakeComponent({
        inlineAttrs: 'bg="blue.200"',
        theme: {
          ...defaultTheme,
          baseStyles: {
            FakeComponent: {
              bg: 'red.400',
              color: 'gray.200'
            }
          }
        }
      })
      expect(asFragment()).toMatchSnapshot()
    })

    it('should accept baseStyles as a function', () => {
      const theme = {
        ...defaultTheme,
        baseStyles: {
          FakeComponent: ({ colorMode, theme }) => ({
            bg: colorMode === 'light' ? 'red.400' : 'red.200',
            color: colorMode === 'light' ? 'black' : 'white',
            border: `2px solid ${
              colorMode === 'light'
                ? theme.colors.blue[600]
                : theme.colors.blue[200]
            }`
          })
        }
      }

      // light mode
      const lightWrapper = renderFakeComponent({
        theme,
        colorMode: 'light',
        inlineAttrs: 'bg="blue.200"'
      })
      expect(lightWrapper.asFragment()).toMatchSnapshot()

      // dark mode
      const darkWrapper = renderFakeComponent({
        theme,
        colorMode: 'dark',
        inlineAttrs: 'bg="vue.200"'
      })
      expect(darkWrapper.asFragment()).toMatchSnapshot()
    })
  })
})
