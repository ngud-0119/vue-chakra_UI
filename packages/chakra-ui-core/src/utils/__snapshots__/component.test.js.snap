// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyledAttrsMixin baseStyles should accept baseStyles as a function 1`] = `
<DocumentFragment>
  .emotion-0 {
  background: var(--chakra-colors-blue-200);
  color: var(--chakra-colors-black);
  border: 2px solid #0045b4;
}

<div
    class="emotion-0"
    data-chakra-component="FakeComponent"
  >
    Fake component
  </div>
</DocumentFragment>
`;

exports[`createStyledAttrsMixin baseStyles should accept baseStyles as a function 2`] = `
<DocumentFragment>
  .emotion-0 {
  background: var(--chakra-colors-vue-200);
  color: var(--chakra-colors-white);
  border: 2px solid #7db1ff;
}

<div
    class="emotion-0"
    data-chakra-component="FakeComponent"
  >
    Fake component
  </div>
</DocumentFragment>
`;

exports[`createStyledAttrsMixin baseStyles should be overiden by props 1`] = `
<DocumentFragment>
  .emotion-0 {
  background: var(--chakra-colors-blue-200);
  color: var(--chakra-colors-gray-200);
}

<div
    class="emotion-0"
    data-chakra-component="FakeComponent"
  >
    Fake component
  </div>
</DocumentFragment>
`;

exports[`createStyledAttrsMixin baseStyles should use theme.baseStyles if given 1`] = `
<DocumentFragment>
  .emotion-0 {
  background: var(--chakra-colors-red-400);
  color: var(--chakra-colors-gray-200);
}

<div
    class="emotion-0"
    data-chakra-component="FakeComponent"
  >
    Fake component
  </div>
</DocumentFragment>
`;
